builder/dockerfile/builder.go

```diff
@@ -196,6 +199,11 @@ func (b *Builder) dispatchDockerfileWithCancellation(dockerfile *parser.Result)
 		// TODO: get this from dispatch
 		imageID = b.image
 
+		if b.Aux != nil {
+			if err := b.Aux.Emit(types.BuildResult{ID: imageID}); err != nil {
+				return "", err
+			}
+		}
```

Comment:
Doing each stage looks like it might be moderately complex (or I've misread the code).

It looks like _before_ calling b.dispatch I need to check `n > 0 && command.From == n.Value` (`imageID != ""` might be better than `n > 0` though) and if true emit the current `imageID` (which is the ID of the layer stage before the `FROM`, i.e. the final layer of the previous stage).

I think I also need to check _after_ calling `b.dispatch` for `n == total` (or possibly `total-1`, need to check the fencepost) and Emit then too to catch the final layer, which naturally won't be followed by a FROM.